subsys/bluetooth/services/ots_olcp.c

```diff
@@ -0,0 +1,311 @@
+	struct bt_gatt_ots *ots, struct bt_gatt_ots_olcp_proc *proc)
+{
+	LOG_DBG("Executing OLCP procedure with 0x%02X Op Code", proc->type);
+
+	switch (proc->type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+		return olcp_first_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+		return olcp_last_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+		return olcp_prev_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+		return olcp_next_proc_execute(ots);
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+		return olcp_goto_proc_execute(ots, proc->goto_params.id);
+	case BT_GATT_OTS_OLCP_PROC_ORDER:
+	case BT_GATT_OTS_OLCP_PROC_REQ_NUM_OBJS:
+	case BT_GATT_OTS_OLCP_PROC_CLEAR_MARKING:
+		return BT_GATT_OTS_OLCP_RES_PROC_NOT_SUP;
+	default:
+		return BT_GATT_OTS_OLCP_RES_INVALID_PARAMETER;
+	}
+};
+
+static enum bt_gatt_ots_olcp_res_code olcp_command_decode(
+	const u8_t *buf, struct bt_gatt_ots_olcp_proc *proc)
+{
+	memset(proc, 0, sizeof(*proc));
+
+	proc->type = *buf++;
+
+	switch (proc->type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+		break;
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+	{
+		struct bt_gatt_ots_obj_id *id;
+
+		id = (struct bt_gatt_ots_obj_id *) &proc->goto_params.id;
+		memcpy(id->ba, buf, sizeof(id->ba));
+		break;
+	}
+	default:
+		LOG_WRN("OLCP unsupported procedure type");
+		return BT_GATT_OTS_OLCP_RES_PROC_NOT_SUP;
+	}
+
+	return BT_GATT_OTS_OLCP_RES_SUCCESS;
+}
+
+static bool olcp_command_len_verify(enum bt_gatt_ots_olcp_proc_type type,
+				    u16_t len)
+{
+	u16_t ref_len = sizeof(type);
+
+	switch (type) {
+	case BT_GATT_OTS_OLCP_PROC_FIRST:
+	case BT_GATT_OTS_OLCP_PROC_LAST:
+	case BT_GATT_OTS_OLCP_PROC_PREV:
+	case BT_GATT_OTS_OLCP_PROC_NEXT:
+	case BT_GATT_OTS_OLCP_PROC_REQ_NUM_OBJS:
+	case BT_GATT_OTS_OLCP_PROC_CLEAR_MARKING:
+		break;
+	case BT_GATT_OTS_OLCP_PROC_GOTO:
+		ref_len += sizeof(((struct bt_gatt_ots_obj_id *) 0)->ba);
+		break;
+	case BT_GATT_OTS_OLCP_PROC_ORDER:
+		ref_len += sizeof(uint8_t);
+		break;
+	default:
+		return true;
+	}
+
+	return (len == ref_len);
+}
+
+static void olcp_ind_cb(struct bt_conn *conn,
+		const struct bt_gatt_attr *attr,
+		u8_t err)
+{
+	LOG_DBG("Received OLCP Indication ACK with status: 0x%04X", err);
+}
+
+static int olcp_ind_send(const struct bt_gatt_attr *olcp_attr,
+			 enum bt_gatt_ots_olcp_proc_type req_op_code,
+			 enum bt_gatt_ots_olcp_res_code olcp_status)
+{
+	u8_t olcp_res[OLCP_RES_MAX_SIZE];
+	u16_t olcp_res_len = 0;
+	struct bt_gatt_ots *ots = (struct bt_gatt_ots *) olcp_attr->user_data;
+
+	/* Encode OLCP Response */
+	olcp_res[olcp_res_len++] = BT_GATT_OTS_OLCP_PROC_RESP;
+	olcp_res[olcp_res_len++] = req_op_code;
+	olcp_res[olcp_res_len++] = olcp_status;
+
+	/* Prepare indication parameters */
+	memset(&ots->olcp_ind.params, 0, sizeof(ots->olcp_ind.params));
+	memcpy(&ots->olcp_ind.attr, olcp_attr, sizeof(ots->olcp_ind.attr));
+	ots->olcp_ind.params.attr = olcp_attr;
```

Comment:
It's also possible to indicate by UUID, rather than keeping the attribute in the OTS instance.

Reply:
Well, it is, but then it's not possible to preserve OTS instance context when choosing UUID option. Instance context is not needed for OLCP callback ``olcp_ind_cb`` for now. However, the same approach is used for OACP where the context is actually used. I would leave it this way for consistency.

This could also be more future proof if we decide to support multiple instances of OTS.

Reply:
I see what you mean. My approach have been to do a lookup (for loop) to find the OTS instance from an `attr` pointer, based on the `service_p` pointer. This way may be better.